File: flink/flink1.14.6/src/main/java/com/flink/java/core/WordCount.java
package com.flink.java.core;

import org.apache.flink.api.common.functions.FlatMapFunction;
import org.apache.flink.api.java.tuple.Tuple2;
import org.apache.flink.streaming.api.datastream.DataStream;
import org.apache.flink.streaming.api.datastream.DataStreamSource;
import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
import org.apache.flink.util.Collector;

public class WordCount {

    public static void main(String[] args) throws Exception {

        // 机器启动 nc -lk 8888
        // 没有nc 需要 yum -y install nc安装

        StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();

        DataStreamSource<String> streamSource = env.socketTextStream("node2.azkaban.bigdata.hw.com", 8888);

        streamSource
                .flatMap(new FlatMapFunction<String, Tuple2<String, Integer>>() {
                    @Override
                    public void flatMap(String s, Collector<Tuple2<String, Integer>> out) throws Exception {

                        //对每一行数据按','切分，切分后的每一个单词输出(word,1)二元组
                        for(String word: s.split(",")){

                            out.collect(new Tuple2<String, Integer>(word, 1));

                        }

                    }
                })
                .keyBy(value->value.f0)
                .sum(1)
                .print();

        env.execute("WordCount");



    }
}

File: flink/flink1.14.6/src/main/scala/com/flink/scala/core/WordCount.scala
package com.flink.scala.core


//导入隐式转换

import org.apache.flink.streaming.api.scala._

object WordCount {


  def main(args: Array[String]): Unit = {


    /**
      * 创建flink环境
      *
      */

    val env: StreamExecutionEnvironment = StreamExecutionEnvironment.getExecutionEnvironment

    //设置并行度
    //默认并行度由系统核数决定
//    env.setParallelism(2)

    //读取socket 构建DS
    //nc -lk 8888
    val linesDS: DataStream[String] = env.socketTextStream("node2.azkaban.bigdata.hw.com",8888)


    //1、将单词拆分
    val wordDS: DataStream[String] = linesDS.flatMap(_.split(","))


    //2、转换成kv格式
    val kvDS: DataStream[(String, Int)] = wordDS.map((_, 1))


    //3、按照key进行分组， 底层也是hash分区
    val keyByDS: KeyedStream[(String, Int), String] = kvDS.keyBy(_._1)

    /**
      * sum算子内部是有状态计算，累加统计
      */
    //4、对value进行聚合
    val countDS: DataStream[(String, Int)] = keyByDS.sum(1)


    //打印结果
    countDS.print()


    //启动flink程序
    env.execute()


  }

}
